File: server/config/callout-types.ts
/**
 * Whitelist of Obsidian callout types that the Markdown parser treats as review questions.
 *
 * This is the SINGLE SOURCE OF TRUTH for question types. It must stay in sync with the
 * `type` CHECK constraint on the `questions` table (supabase/migrations/0002_tables.sql).
 * The parser (added in a later pass) reads `> [!<type>]` callouts and keeps only the ones
 * whose type is in QUESTION_TYPES.
 */
export const QUESTION_TYPES = ['recall', 'scenario', 'compare', 'explain'] as const;

export type QuestionType = (typeof QUESTION_TYPES)[number];

/** Type guard: is `value` one of the whitelisted question types? */
export function isQuestionType(value: string): value is QuestionType {
  return (QUESTION_TYPES as readonly string[]).includes(value);
}

File: backend/app/config/callout_types.py
"""Single source of truth for review question types.

Must stay in sync with the `type` CHECK constraint on the `questions` table
(supabase/migrations/0002_tables.sql). Questions are AI-generated per user from
the lesson content (sub-project #4a); this is the enum the model must choose
from when producing structured output, plus a Vietnamese description of each
type used to steer that choice inside the prompt.
"""

QUESTION_TYPES: list[str] = ["recall", "scenario", "compare", "explain"]

# Shown to the model so it picks a fitting type per question. Keys must match
# QUESTION_TYPES exactly.
QUESTION_TYPE_DESCRIPTIONS: dict[str, str] = {
    "recall": "Nhớ lại một sự kiện, định nghĩa hoặc chi tiết cụ thể trong bài.",
    "scenario": "Áp dụng kiến thức của bài vào một tình huống thực tế mới.",
    "compare": "So sánh hoặc đối chiếu hai khái niệm, cách tiếp cận trong bài.",
    "explain": "Giải thích nguyên nhân, cơ chế hoặc lý do đằng sau một ý trong bài.",
}
